### PR TITLE
Ddc new project

### PR DESCRIPTION
**Pull request by devoncarew**
- add a `sky_tools` binary. This supports the `init` command, used for creating new sky projects.
- create a single project type. This is a very simple, material design hello world project
- add a test to make sure that the project we create is well formed (has a valid pubspec and the dart code analyses well.

```shell
pub run sky_tools init -ofoobar

Creating foobar...
foobar/.gitignore
foobar/pubspec.yaml
foobar/README.md
foobar/lib/main.dart

Running pub get...
Resolving dependencies...
+ args 0.13.2
+ cassowary 0.1.7
...
Precompiled sky:init.
Precompiled sky_tools:build_sky_apk.
Precompiled sky_tools:sky_server.

All done! To run your application:
cd foobar
./packages/sky/sky_tool start
```

### CODE REVIEW

**Comment by abarth** [comment]
What is "Ddc" ?

**Comment by devoncarew** [comment]
That's me :) I'm in the habit of prefixing my branches with my initials, and `ddc_new_project` gets converted into title case my github at some point in the process...

File: lib/src/init.dart [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+class InitCommandHandler extends CommandHandler {
+  InitCommandHandler() : super('init', 'Create a new sky project.');
+
+  ArgParser get parser {
+    ArgParser parser = new ArgParser();
+    parser.addFlag('help',
+        abbr: 'h', negatable: false, help: 'Display this help message.');
+    // parser.addOption('template',
+    //     abbr: 't',
+    //     help: "The template to generate; either sky-simple or sky-full.");
```
Comment (abarth) [inline lib/src/init.dart]: Please don't check in commented out code.  :)

File: lib/src/init.dart [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    if (!results.wasParsed('out')) {
+      printUsage('No option specified for the output directory.');
+      return new Future.value();
+    }
+
+    Directory out = new Directory(results['out']);
+
+    // TODO: confirm overwrite with user
+    //if (out.existsSync())
```
Comment (abarth) [inline lib/src/init.dart]: Ditto

File: lib/src/init.dart [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  void generateInto(Directory dir) {
+    String projectName = _normalizeProjectName(p.basename(dir.path));
+    print('Creating ${p.basename(projectName)}...');
+    dir.createSync(recursive: true);
+
+    files.forEach((String path, String contents) {
+      contents = contents
+          .replaceAll('{{projectName}}', projectName)
+          .replaceAll('{{description}}', description);
```
Comment (abarth) [inline lib/src/init.dart]: Is there not a real templating system in Dart we should use?  In Python, I would recommend jinja2...

**Comment by abarth** [comment]
LGTM except for the hand-rolled templating.  I'd prefer if we used a real templating system.

**Comment by abarth** [comment]
Is https://pub.dartlang.org/packages/mustache any good?

File: lib/src/init.dart (hunk 1 above)
Comment (devoncarew) [inline lib/src/init.dart]: done

File: lib/src/init.dart (hunk 2 above)
Comment (devoncarew) [inline lib/src/init.dart]: done

File: lib/src/init.dart (hunk 3 above)
Comment (devoncarew) [inline lib/src/init.dart]: We use `mustache4dart` in dartdoc. Happy to use either one. Longer term I see the sky templates being stagehand plugins, and `sky_tools` being a wrapper around that.

@sethladd and I need to figure out the story for plugins into stagehand though, so putting some project creation smarts into `sky_tools` seems like a good solution for now.

I'll do some quick investigation of `mustache` vs `mustache4dart`.

**Comment by devoncarew** [comment]
@abarth, I updated this PR to use mustache4dart. From talking w/ @sethladd, I think we want to push on getting support for 3rd party plugins into stagehand to support use cases just like this. So I'd like to:
- land this PR
- publish a new version of the package
- update atom to create sky projects using this tool
- revisit stagehand's plugin story
- update this tool so that the project creation portion is re-hosted on stagehand

Sound good?

**Comment by abarth** [comment]
Sounds great

**Comment by abarth** [comment]
Lgtm

**Pull request merged**